pkg/virt-handler/device-manager/device_controller.go

```diff
@@ -233,38 +272,56 @@ func (c *DeviceController) refreshPermittedDevices() {
 	// This function can be called multiple times in parallel, either because of multiple
 	//   informer callbacks for the same event, or because the configmap was quickly updated
 	//   multiple times in a row. To avoid starting/stopping device plugins multiple times,
-	//   we need to protect c.devicePlugins, which we read from in
+	//   we need to protect c.startedPlugins, which we read from in
 	//   c.updatePermittedHostDevicePlugins() and write to below.
-	c.devicePluginsMutex.Lock()
+	c.startedPluginsMutex.Lock()
+	defer c.startedPluginsMutex.Unlock()
 
 	enabledDevicePlugins, disabledDevicePlugins := c.updatePermittedHostDevicePlugins()
 
 	// start device plugin for newly permitted devices
 	for resourceName, dev := range enabledDevicePlugins {
-		go c.startDevicePlugin(dev)
-		c.devicePlugins[resourceName] = dev
+		c.addAndStartDevice(resourceName, dev)
 		debugDevAdded = append(debugDevAdded, resourceName)
 	}
 	// remove device plugin for now forbidden devices
-	for resourceName, dev := range disabledDevicePlugins {
-		close(dev.stopChan)
-		delete(c.devicePlugins, resourceName)
+	for resourceName, _ := range disabledDevicePlugins {
+		c.removeAndStopDevice(resourceName)
 		debugDevRemoved = append(debugDevRemoved, resourceName)
 	}
 
-	c.devicePluginsMutex.Unlock()
-
 	logger.Info("refreshed device plugins for permitted/forbidden host devices")
 	logger.Infof("enabled device-plugins for: %v", debugDevAdded)
 	logger.Infof("disabled device-plugins for: %v", debugDevRemoved)
 }
 
+func (c *DeviceController) addAndStartDevice(resourceName string, dev controlledDevice) {
+	c.removeAndStopDevice(resourceName)
+	dev.backoff = c.backoff
+	dev.Start()
+	c.startedPlugins[resourceName] = dev
+}
+
+func (c *DeviceController) removeAndStopDevice(resourceName string) {
+	dev, exists := c.startedPlugins[resourceName]
+	if exists {
+		dev.Stop()
+		delete(c.startedPlugins, resourceName)
+	}
+}
+
 func (c *DeviceController) Run(stop chan struct{}) error {
 	logger := log.DefaultLogger()
+
 	// start the permanent DevicePlugins
-	for _, dev := range c.devicePlugins {
-		go c.startDevicePlugin(dev)
-	}
+	func() {
+		c.startedPluginsMutex.Lock()
+		defer c.startedPluginsMutex.Unlock()
+		for name, dev := range c.permanentPlugins {
+			c.addAndStartDevice(name, dev)
+		}
+	}()
```

Comment:
nit: would it make sense to export a function to execute a piece of code within the device controller's locked mutex scope ? 

Something like:
```golang
func (c *DeviceController) executeCriticalSection(f func()) {
    c.startedPluginsMutex.Lock()
    defer c.startedPluginsMutex.Unlock()
    f()
}

// callers
c.executeCriticalSection(func() {
    for name, dev := range c.permanentPlugins {
        c.addAndStartDevice(name, dev)
    }
})
```